Defining semantics of instructions unsupported in Strata/Stoke.

sdasgup3 opened the pull request:
This includes the support for memory instructions and packed instructions with arity 3.

All the defined [instructions](https://github.com/sdasgup3/stoke/blob/improve_instruction_support/artifacts/extra_support_needs_tobe_added.txt) have been manually defined and tested using the following strategies:

1. Test the porting: All these instruction semantics had been developed in a
branch of Stoke which had been used for the Strata project (henceforth referred
    as Strata's Stoke branch). So, I have to port all those semantics to the
latest `develop` branch of the upstream Stoke. In order to test that porting, I
matched the Z3 formula, using usat check,  of all the CPU states relevant for
those instructions.   All the equivalent tests can be found at
[link](https://github.com/sdasgup3/stoke/tree/improve_instruction_support/artifacts/equivTests).
In the link, the *.strata.Z3.sym represent the Z3 formula obtained from
Strata's Stoke branch and *.stoke.Z3.sym represent that obtained from the
'develop' branch of latest Stoke. *.certificate.z3 represents the unsat check
expression, created using the above two Z3 expressions, that needed to be solved for equivalence.

2. Test the instruction behaviour using co-simulation against hardware:  The
test-inputs used are the public available 6630 test-cases used for the Strata
project.

sdasgup3 committed:
Defining semantics of instructions unsupported in Strata/Stoke.
This includes the support for packed instructions with arity 3.

All the defined
[instructions](https://github.com/sdasgup3/stoke/blob/improve_instruction_support/artifacts/extra_support_needs_tobe_added.txt)
have been manually defined and tested using the following strategies:

1. Test the porting: All these instruction semantics had been developed in a
branch of Stoke which had been used for the Strata project (henceforth referred
    as Strata's Stoke branch). So, I have to port all those semantics to the
latest `develop` branch of the upstream Stoke. In order to test that porting, I
matched the Z3 formula, using usat check,  of all the CPU states relevant for
those instructions.   All the equivalent tests can be found at
[link](https://github.com/sdasgup3/stoke/tree/improve_instruction_support/artifacts/equivTests).
In the link, the *.strata.Z3.sym represent the Z3 formula obtained from
Strata's Stoke branch and *.stoke.Z3.sym represent that obtained from the
'develop' branch of latest Stoke. *.certificate.z3 represents the unsat check
expression, created using the above two Z3 expressions, that needed to be solved for equivalence.

2. Test the instruction behaviour using co-simulation against hardware:  The
test-inputs used are the public available 6630 test-cases used for the Strata
project.

sdasgup3 committed:
Defining semantics of instructions unsupported in Strata/Stoke.
This includes the support for packed instructions with arity 3.

All the defined
[instructions](https://github.com/sdasgup3/stoke/blob/improve_instruction_support/artifacts/extra_support_needs_tobe_added.txt)
have been manually defined and tested using the following strategies:

1. Test the porting: All these instruction semantics had been developed in a
branch of Stoke which had been used for the Strata project (henceforth referred
    as Strata's Stoke branch). So, I have to port all those semantics to the
latest `develop` branch of the upstream Stoke. In order to test that porting, I
matched the Z3 formula, using usat check,  of all the CPU states relevant for
those instructions.   All the equivalent tests can be found at
[link](https://github.com/sdasgup3/stoke/tree/improve_instruction_support/artifacts/equivTests).
In the link, the *.strata.Z3.sym represent the Z3 formula obtained from
Strata's Stoke branch and *.stoke.Z3.sym represent that obtained from the
'develop' branch of latest Stoke. *.certificate.z3 represents the unsat check
expression, created using the above two Z3 expressions, that needed to be solved for equivalence.

2. Test the instruction behaviour using co-simulation against hardware:  The
test-inputs used are the public available 6630 test-cases used for the Strata
project.

bchurchill commented:
It looks like the CI failure doesn't involve your code.  I'll troubleshoot when I get a chance.  Your PRs are the first with this CI system!

sdasgup3 committed:
Fuzz tested the supported instructions.

sdasgup3 committed:
Fuzz tested the supported instructions.

sdasgup3 committed:
Minor fix to allow fuzz testing the added instructions

sdasgup3 committed:
Minor fix to allow fuzz testing the added instructions

sdasgup3 commented:
@bchurchill  , Thanks for the update

sdasgup3 committed:
Fixing the cmpxchg implementation based on a bug caught during the fuzz testing.
The bug appears when the memory effectve address involves %rax, which is same as
the accumulator to be modifed by the instruction. The ordering of destination update
and accumulator update needs to be corrected to avoid reads from a location specified by
the updated value of %rax.

sdasgup3 committed:
Fixing the cmpxchg implementation based on a bug caught during the fuzz testing.
The bug appears when the memory effectve address involves %rax, which is same as
the accumulator to be modifed by the instruction. The ordering of destination update
and accumulator update needs to be corrected to avoid reads from a location specified by
the updated value of %rax.

Commit:
fixing Makefile for #998

Commit:
wip on creating base image

bchurchill committed:
new attempt at building stoke automatically

bchurchill committed:
simplifying travis script

bchurchill committed:
Installing upgrades when building docker image

bchurchill committed:
reorganizing docker-related scripts into folder

bchurchill committed:
docker image should not already have built code; updating README

bchurchill committed:
only try to push if it's not a pull request

bchurchill committed:
Merge branch 'develop' into issue-998

bchurchill committed:
travis.yml bug fix

bchurchill committed:
Merge branch 'issue-998' of github.com:StanfordPL/stoke into issue-998

bchurchill committed:
Merge pull request #999 from StanfordPL/issue-998

Setting up travis-ci for pull requests.  Closes #998.

bchurchill commented:
Please try merging develop branch into your pull request.  travis-ci should build it and hopefully it will work.

sdasgup3 committed:
Merge branch 'improve_instruction_support_PR1' of github.com:sdasgup3/stoke into improve_instruction_support_PR1
